rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,152 @@
+mechanism.
+
+The CI version linter currently compares changes to versioned types
+against the PR base branch. That's more fine-grained than needed, if
+the goal of preventing version changes is to allow interoperability
+with deployed software. Instead, compare the types against the branch 
+of the latest released software. That strategy will reduce the amount
+of noise produced by the linter, and provide more useful results.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Version-tagging changes
+
+Currently, tagging a `Stable` module with `%versioned` produces
+shadowing `bin_read_t` and `bin_write_t` functions that read and write
+version tags, positive integers, prepended to serialized data, via the
+`ppx_version` library. The proposal is not to shadow those functions,
+and instead use those generated by `deriving bin_io`, directly.  It is
+still useful to use `%versioned` to assure that serializations do not
+change for a given versioned type.
+
+For backwards-compatibility, we can add the annotation
+```
+ [@@@with_all_version_tags]
+```
+to a `Stable` module. For each versioned type module `Vn`, generate a
+module:
+```ocaml
+ module With_all_version_tags = struct
+   type t = ...
+   
+   let bin_read_t = ... bin_read_t ...
+   let bin_write_t = ... bin_write_t ...
+   ...
+ end
+```
+where the shadowing functions are generated as they are today. For
+this case, the constituent types must also have the same annotation,
+which is enforced by the construction of the type `t`: the constituent
+types of `t` are of the form `M.Stable.Vn.With_all_version_tags.t`.
+
+Another available annotation can be
+```
+ [@@@with_top_version_tag]
+```
+to generate:
+```ocaml
+ module With_top_version_tag = struct
+   type t = ...
+   
+   let bin_read_t = ... bin_read_t ...
+   let bin_write_t = ... bin_write_t ...
+   ...
+ end
+```
+where the shadowing functions handle a single version tag to the
+serializations of outermost type instances. The constituent types need
+not have the annotation: the type `t` will be identical to the type
+`t` in the containing `Vn` module.
+
+The existing versioning system generates a function in `Stable` modules:
+```ocaml
+ val bin_read_to_latest_opt : Bin_prot.Common.buf -> pos_ref:(int ref) -> Stable.Latest.t option
+```
+That function deserializes data by dispatching on version tags.
+Because the default serialization does not contain version tags, it no
+longer makes sense to generate such a function in `Stable`. Instead,
+generate it inside `With_all_version_tags` and `With_top_version_tag`,
+if they're created.
+
+### Version linting changes
+
+The CI version linter looks for changes to versioned types in a PR
+branch by comparing those types with counterparts in the base
+branch. The linter will complain if a new version is added, and later
+modified, even if there are no deployed nodes using the new version.
+
+Instead, the linter should use a release branch for comparison. It
```

Comment:
Will this comparison continue to be done on the OCaml type definition, or can we modify this to operate off of the `bin_shape` representation instead? We had discussed this some in the past, and I think utilizing the `bin_shape` is more flexible as we can modify the OCaml type definition at will and provide custom `bin_prot` definitions to maintain compatibility.

Reply:
Currently, the comparison is done on the textual representation of canonicalized types.

We can instead do the comparison on the `Shape.eval_to_digest_string bin_shape_t`, which will allow custom serializations.

In CI, we'd report the differing types, when they do differ, and the digests when the types haven't changed.

Reply:
To support custom serializers fully, we'd need something like `%versioned_custom`, similar to `%versioned_binable`, where we omit `deriving bin_io`.